### ISSUE TITLE
Named-argument taint stop-gap: key the variadic rule on name, handle late static binding, and survive re-entrant file analysis

### ISSUE THREAD

**Issue by alies-dev** [psalm/laravel-psalm-plugin#1406]
Follow-ups deferred from #1396 (the stop-gap for upstream vimeo/psalm#11923, which fixes #1395).

#1396 ships a deliberately minimal core: strip taint from a named argument's value unless the callee is statically resolvable and the declared parameter at the written offset already matches the argument's name, plus an exclusion for nodes that are themselves sink subjects (`Eval_`, `Include_`, dynamic `FuncCall`, dynamic `New_`).

Attempts to widen it past that produced regressions in both directions, so the wider work was cut and is recorded here. All of it is moot if vimeo/psalm#11923 is fixed upstream, which retires the handler entirely.

## 1. The variadic rule must key on name, not written offset

PHP matches a named argument by **name**; the handler reasons about `params[writtenOffset]`. One root cause, two opposite symptoms:

```php
// (a) wrongly PRESERVED -> false positive
/** @psalm-taint-sink html $values */
function vsink(string $target = '', string ...$values): void { foreach ($values as $v) { echo $v; } }
vsink(extra: 'safe', target: tainted());
// runtime: $target is tainted, $values === ['extra' => 'safe']; no html exposure.
// The name matches the EARLIER $target, but written offset 1 collides with the variadic's
// declared offset 1, so Psalm emits a false TaintedHtml.

// (b) wrongly STRIPPED -> false negative
/** @psalm-taint-sink file $a
 *  @psalm-taint-sink html $rest */
function v(string $a = '', string $b = '', string ...$rest): void { echo $a; echo $b; foreach ($rest as $r) { echo $r; } }
v(zzz: tainted());
// An unmatched name lands in the variadic wherever it is written; offset 0 points at $a,
// which is not variadic, so the genuine finding is stripped.
```

A correct fix mirrors `ArgumentsAnalyzer`'s own matching: find the parameter whose name matches, and fall back to the variadic only when none does.

## 2. `static::` and late static binding never consult the child's signature

```php
abstract class P {
    public function go(): void { static::sink(label: tainted()); }
    /** @psalm-taint-sink file $path */
    public static function sink(string $path = '', string $label = ''): void { echo $path; }
}
final class C1 extends P {}
final class C2 extends P {
    /** @psalm-taint-sink html $label */
    public static function sink(string $label = '', string $path = ''): void { echo $label; }
}
```

Vanilla reports `2 TaintedHtml + 2 TaintedTextWithQuotes`; the plugin reports nothing. `resolveClassNamePart()` maps `static` to `Context::self`, which is the **declaring** class in every pass, so an overriding child signature is never seen and no pass can heal the record. Traits avoid this only because `self` differs per consuming class.

## 3. Re-entrant `beforeFileAnalysis` drops pending records

```php
sink(setup: include 'setup.php', path: tainted());
```

The nested include recursively enters `FileAnalyzer`, whose `BeforeFileAnalysis` replaces the per-file map, so the later `path:` record is gone by the time it is read. Reproduced. A correct fix needs to tell a root file's analysis start from a nested file entered mid-analysis, which `BeforeFileAnalysisEvent` does not expose cleanly.

## 4. Accepted false negatives, documented in the handler

Not bugs to fix independently of the above, but they should be revisited together:

- named arguments on `MethodCall` / `NullsafeMethodCall` are stripped unconditionally (the receiver's type is unresolved at `BeforeExpressionAnalysis` time);
- a named argument captured by a variadic that carries its own sink is stripped;
- `TaintKind::ALL_INPUT` excludes `USER_SECRET` / `SYSTEM_SECRET`, so secret-taint mis-attribution survives the strip.

## Prevalence

A psalm-delta run over the 18-app benchmark corpus showed **zero** movement in taint findings for the shipped core, so none of these shapes occur on a sink anywhere in that corpus. Worth weighing before investing in the name-based rewrite.

**Comment by alies-dev** [comment]
Scope correction after #1413.

**Removed from this issue: the facade gap.** Item 4's "named arguments on `MethodCall` / `NullsafeMethodCall` are stripped unconditionally" understated the problem. `Codebase::getFunctionLikeStorage()` gates on `methodExists()` with `$with_pseudo` false, so no facade `@method static` pseudo-method ever resolved either, and every facade `StaticCall` with a named argument was stripped. Measured by disabling only the handler registration:

| fixture | handler off | `abaccc94` |
|---|---|---|
| `File::delete(paths: tainted())` | `TaintedFile` | silent |
| `Storage::get(path: tainted())` | `TaintedFile` | silent |
| `$fs->delete(paths: tainted())` | `TaintedFile` | silent |

A post-population sweep of `pseudo_static_methods` for `sinks != 0` finds 194 entries across 97 distinct methods (File 36, Storage 28, Http 9, Response 8, View 6, Redirect 5, Process 3, Artisan 2, canonical FQCN and root alias each). Fixed in #1413.

**Corrected: the prevalence claim.** The "zero movement across the 18-app corpus" line at the bottom of this issue, and the matching paragraph the handler docblock used to carry, are both wrong for the same reason. That scan resolved method names through real method storage, the same lookup that misses pseudo-methods, so it was structurally blind to the class it was meant to rule out. The shape does occur in the public benchmark set: `vito` calls `File::makeDirectory(path: dirname($location), recursive: true)` and `File::ensureDirectoryExists(path: dirname($extractPath))`, and both carry `@psalm-taint-sink file $path`.

**Narrowed: item 4's method-call bullet.** A receiver that is a plain `$var` holding exactly one `TNamedObject` now resolves from `Context::vars_in_scope`, which covers `$this->m()` and every DI-injected service. What remains is a receiver that is not a variable.

**Added: the chained receiver.** `Storage::disk('local')->put(path: $input)` is an expression, not a variable, so there is no entry to read and no receiver type available at the pre-pass. It still strips while the equivalent two-line variable form does not, which is a surprising split for a reader. Prevalence across the 17 public benchmark apps is zero. Executed and confirmed on `Storage::disk()->put`, `Http::withToken()->get`, `DB::connection()->select` and `DB::table()->whereRaw`.

**Unchanged:** items 1 (variadic keyed on offset, not name), 2 (`static::` and late static binding), 3 (re-entrant `beforeFileAnalysis`), and item 4's remaining bullets. Item 2 is now the only late-static-binding hole: the plain-variable receiver path is not affected, because `vars_in_scope` holds the same type `MethodCallAnalyzer` resolves the call against, so the two cannot disagree.

**Comment by alies-dev** [comment]
Follow-up to the previous comment, adding the residual list in full.

### Accepted false-negative surface after #1413

**1. Facade `@method static` pseudo-methods. CLOSED.** Covered above.

**2. `MethodCall` / `NullsafeMethodCall`. Narrowed, not closed.** A receiver that is a plain `$var` already in scope now resolves from `Context::vars_in_scope` when its type is a single `TNamedObject`, which covers `$this->m()` and every DI-injected service. Still stripped:

* A chained or fluent receiver, because it is an expression rather than a `Variable` and therefore has no `vars_in_scope` entry: `Storage::disk('local')->put(path: $x, contents: $y)`, `Http::withToken($t)->get(url: $u)`, `DB::connection()->select(query: $q)`, `DB::table('users')->whereRaw(sql: $s)`. All four are silent while the positional form reports. No receiver type exists at `beforeExpressionAnalysis` time, so this cannot be fixed inside the pre-pass. Assigning the receiver to a variable first restores detection.
* A union or non-object receiver, deliberately: narrowing on anything but exactly one known class turns into false positives.
* Any other non-`Variable` receiver, for example `(new Sink())->m(...)`, `$arr[0]->m(...)`, `$this->prop->m(...)`.

**3. Variadic capture.** Unchanged.
**4. `static::` and `self::` late static binding.** Unchanged, and now the only late-static-binding hole. The plain-variable receiver path is not affected: `vars_in_scope` holds the same type `MethodCallAnalyzer` resolves the call against, so the two cannot disagree.
**5. Re-entrant `beforeFileAnalysis`.** Unchanged.
**6. Callees that stay unresolvable** (a dynamic function or class expression, an internal name that misses both storage and the CallMap). Unchanged.

### Prevalence of the chained-receiver residual: zero

A scan of all 17 public benchmark apps for `->{sink method name}(name:` across the File, Storage, Http, View, DB, Response, Redirect and Process sink method names returned no sites.

### How this was measured

A facade call has no vanilla-Psalm counterpart to diff against, so the oracle is byte equality of the issue list on one fixture with only this handler's effect disabled. Corpus silence is not evidence of soundness: the corpus scan and three review rounds all reported zero movement while the facade gap was live.
